lib/js_of_ocaml/typed_array.mli

```diff
@@ -51,41 +53,58 @@ class type ['a, 'b] typedArray = object
 
   method set_fromArray : 'a js_array t -> int -> unit meth
 
-  method set_fromTypedArray : ('a, 'b) typedArray t -> int -> unit meth
+  method set_fromTypedArray : ('a, 'b, 'c) typedArray t -> int -> unit meth
 
-  method subarray : int -> int -> ('a, 'b) typedArray t meth
+  method subarray : int -> int -> ('a, 'b, 'c) typedArray t meth
 
-  method subarray_toEnd : int -> ('a, 'b) typedArray t meth
+  method subarray_toEnd : int -> ('a, 'b, 'c) typedArray t meth
 
-  method slice : int -> int -> ('a, 'b) typedArray t meth
+  method slice : int -> int -> ('a, 'b, 'c) typedArray t meth
 
-  method slice_toEnd : int -> ('a, 'b) typedArray t meth
+  method slice_toEnd : int -> ('a, 'b, 'c) typedArray t meth
 
-  method _content_type_ : 'b optdef readonly_prop
+  (* This fake method is needed for typing purposes. Without it, ['b] would not
+     be constrained. *)
+  method _content_type_ : ('b * 'c) optdef readonly_prop
 end
 
-type int8Array = (int, Bigarray.int8_signed_elt) typedArray
+type int8Array = (int, int, Bigarray.int8_signed_elt) typedArray
+
+type uint8Array = (int, int, Bigarray.int8_unsigned_elt) typedArray
 
-type uint8Array = (int, Bigarray.int8_unsigned_elt) typedArray
+type int16Array = (int, int, Bigarray.int16_signed_elt) typedArray
 
-type int16Array = (int, Bigarray.int16_signed_elt) typedArray
+type uint16Array = (int, int, Bigarray.int16_unsigned_elt) typedArray
 
-type uint16Array = (int, Bigarray.int16_unsigned_elt) typedArray
+type int32Array = (int32, Int32.t, Bigarray.int32_elt) typedArray
 
-type int32Array = (int32, Bigarray.int32_elt) typedArray
+type uint32Array = (uint32, Int32.t, Bigarray.int32_elt) typedArray
 
-type uint32Array = (int32, Bigarray.int32_elt) typedArray
+type float32Array = (number_t, float, Bigarray.float32_elt) typedArray
 
-type float32Array = (float, Bigarray.float32_elt) typedArray
+type float64Array = (number_t, float, Bigarray.float64_elt) typedArray
 
-type float64Array = (float, Bigarray.float64_elt) typedArray
+type ('bigarray, 'typed_array, 'elt) kind =
+  | Char : (int, char, Bigarray.int8_unsigned_elt) kind
```

Comment:
Why is it an int here and not a char ?

Reply:
Good question, I may be mistaken but I see no good reason why not (cc @vouillon). I have changed to `char, char` in the last version, and added a test for a `char` bigarray. One quirk is that converting such a bigarray to a typed array will yield a Uint8Array, but I think this is to be expected given that Javascript does not have CharArray.

Reply:
Why is it unsound?

Reply:
This can be used to turn a float into an int, for instance:
```ocaml
type _ t =
  | A : float -> int t
  | B : int -> char t
  | C : float t
  | D : int32 t
  | E : int64 t
  | F : nativeint t
  | G : Complex.t t

let f (type a b) (x : (a, b) Bigarray.kind) : a t =
  match x with
  | Char -> B 10
  | Int8_unsigned -> A 10. | Int8_signed -> A 10.
  | Int16_unsigned -> A 10. | Int16_signed -> A 10. | Int -> A 10.
  | Float32 -> C | Float64 -> C
  | Int32 -> D
  | Int64 -> E
  | Nativeint -> F
  | Complex32 -> G | Complex64 -> G

let g (x : char t) = match x with B n -> n

let x = (Obj.magic Bigarray.Int8_unsigned : (char, Bigarray.int8_unsigned_elt) Bigarray.kind)

let () = Format.printf "%d@." (g (f x))
```

Reply:
Got it. I just pushed a commit to disallow converting from `Char` bigarrays to typed arrays. It seems to me like the natural solution since there are no such TypedArrays, and the previous API also didn’t allow it.